Add a new gas cost model

koute opened the pull request:
[Rendered](https://github.com/koute/new-gas-cost-model/blob/master/graypaper.pdf).

This PR makes the following changes:

* Adds a new gas cost model (see sections A.9 and A.10 in the PVM appendix) which, given a basic block of instructions, calculates the gas cost for the whole block.
* Removes the `sbrk` instruction and replaces it with a hostcall.
* Switches to per-basic-block gas metering. Gas is now charged every time we enter a basic block.
* Renumbers the instruction opcodes and hostcall numbers to account for the new changes.

**I highly recommend reviewing commit-by-commit** (since some of the changes like e.g. renumbering the opcodes result in quite big diffs, but are otherwise trivial).

A few general points about the changes here:

* The new gas cost model was successfully implemented by at least 2 independent JAM teams based on the equations contained within this PR.
* I consider the gas cost algorithm to be *final*, however the cost tables and possibly some of the numeric constants/factors *will* still change as I work on auditing the model.
* The gas cost of memory accesses is currently hardcoded to be always equivalent to the L2 cache access cost; this will be replaced with a proper equation in a future PR, along with an extra mechanism to gas meter the paging mechanism itself (so that we can guarantee an active working set of at most the size of the L3 cache).

koute committed:
Add new gas cost model

koute committed:
Remove gas costs from normal instruction table

koute committed:
Switch to gas metering per basic block

koute committed:
Replace `sbrk` with `grow_heap` hostcall

koute committed:
Renumber opcodes so that there are no holes

koute committed:
Renumber hostcalls

koute committed:
Bump version to 0.8.0

koute committed:
Switch to gas metering per basic block: revert extra arg in call to Psi_H

koute committed:
Lower the per-page gas cost for `grow_heap` hostcall

We need direct metering of the page mapping mechanism
to properly meter the memory anyway, so for now lower
the per-page gas costs for compatibility with already
written JAM binaries so that their gas costs don't
baloon orders of magnitude into the stratosphere.

vekexasia commented:
hey @koute can you tell me what **o** and `s` are suppoused to be in the grow_heap?

<img width="569" height="282" alt="image" src="https://github.com/user-attachments/assets/bf8bd9e5-3921-44c4-9736-d67f177710b0" />

vekexasia commented:
> The hint is in the `\mathbf{p}` being passed to this host call, the same one from "Standard Program Initialization" section. Once you look at that it should be obvious what `o` and `s` are.

That would explain a lot thanks. Just so you know \mathbf{p} needs some disambiguation imho. for example if you look at is authorized you would think p is a totally different thing

<img width="960" height="407" alt="image" src="https://github.com/user-attachments/assets/a28e5d47-a4dc-4915-ad4a-3cea02098377" />

koute committed:
Clean up PVM blob and JAM blob variable names

koute committed:
Make the provenance of `o` and `s` explicit in `grow_heap` hostcall

koute commented:
Okay, I've cleaned up the notation and the terminology for JAM and PVM program blobs (which was somewhat confusing, using the same letter for different-but-closely-related things), and I've made the source of the `o` and `s` in `grow_heap` explicit.

The new notation/terminology is as follows:

- fancy `j` - JAM program blob for a toplevel service, contains a PVM program blob inside, plus the memory map
- fancy `p` - PVM program blob (instruction opcodes, bitmask and jump table); exactly the same blob that is also passed when creating an inner PVM

It turns out that the JAM program blob is readily available in all of the PVM invocation functions, so I didn't have to do any heavy plumbing and it was essentially just an issue of passing in the correct variable to the `grow_heap`'s subfunction, and then extracting what we need inside. No teleportation necessary.

This should be good to go now? (Pending any further reviews, of course.)

davxy commented on text/pvm.tex:
```diff
@@ -124,35 +125,59 @@ \subsection{Basic Blocks and Termination Instructions}
   }
 \end{equation}
 
+On the very first step of execution, and every time the execution enters a new basic block or jumps back to the beginning of the current basic block, the gas counter of the machine is updated according to the gas cost $\gascostforblock{\startofbasicblock(\imath)}$~(\ref{eq:gascostforblock}) of the target basic block. No instruction is allowed to execute within a basic block without first paying t ...
```
```suggestion
On the very first step of execution, and every time the execution enters a new basic block or jumps back to the beginning of the current basic block, the gas counter of the machine is updated according to the gas cost $\gascostforblock{\startofbasicblock(\imath)}$~(\ref{eq:gascostforblock}) of the target basic block. No instruction is allowed to execute within a basic block unless the gas cost for the entire basic block has been charged in advance. In case there's not enough gas remaining to cover the full gas cost, the execution is interrupted and the gas counter remains unchanged. Formally:
```

Just a personal preference to express the same concept, feel free to ignore

davxy commented on text/pvm.tex:
```diff
@@ -427,35 +454,27 @@ \subsubsection{Instructions with Arguments of Two Registers}
+  108&\token{sign\_extend\_16}&$\reg'_D = \unsigned{\signedn{2}{\reg_A \bmod 2^{16}}}$\\ \mrule
+  109&\token{zero\_extend\_16}&$\reg'_D = \reg_A \bmod 2^{16}$\\ \mrule
+  110&\token{reverse\_bytes}&$\forall i \in \N_8 : \encode[8]{\reg'_D}\sub{i} = \encode[8]{\reg_A}_{7-i}$\\
+%  10X&\token{}&$\reg'_D = ...$\\ \mrule
```
```suggestion
```

Dead code?

davxy commented on text/pvm.tex:
```diff
@@ -714,32 +733,33 @@ \subsection{Host Call Definition}
+  &\Psi_H^*\colon \abracegroup{
     \tuple{\begin{aligned}
-      &\blob, \pvmreg, \gas, \regs,\\&\ram, \contextmutator{X}, X
+      &\blob, \pvmreg, \gas, \regs,\\&\ram, \bool, \contextmutator{X}, X
```
```suggestion
      &\blob, \pvmreg, \gas, \bool, \regs,\\&\ram, \contextmutator{X}, X
```

davxy commented on text/pvm_invocations.tex:
```diff
@@ -66,7 +67,7 @@ \subsection{Refine Invocation}\label{sec:refineinvocation}
 
 
-The inner \textsc{pvm} invocation host-calls, meanwhile, depend on an integrated \textsc{pvm} type, which we shall denote $\pvmguest$. It holds some program code, instruction counter and \textsc{ram}:
+The inner \textsc{pvm} invocation host-calls, meanwhile, depend on an integrated \textsc{pvm} type, which we shall denote $\pvmguest$. It holds a PVM program blob, instruction counter and \textsc{ram}:
```
```suggestion
The inner \textsc{pvm} invocation host-calls, meanwhile, depend on an integrated \textsc{pvm} type, which we shall denote $\pvmguest$. It holds a \textsc{pvm} program blob, instruction counter and \textsc{ram}:
```

davxy commented on text/pvm.tex:
```diff
@@ -839,3 +859,413 @@ \subsection{Argument Invocation Definition}
+  \end{cases}
+\end{equation}
+
+Assuming $m \text{ such that } \Xi_{m + 1} = \none$ the final gas cost of a basic block starting at $\imath$ is defined as:
```
This is not clear to me. How is the output of $\Xi_{m + 1} = ∅$?

davxy reviewed (commented):
Overall looks good, just few questions

koute commented on text/pvm.tex:
```diff
@@ -839,3 +859,413 @@ \subsection{Argument Invocation Definition}
+\begin{equation}
+  \Xi_{n + 1}''(\Xi_n) = \abracegroup{
+    \rob¬state^{(n + 1)}_{\find¬readytostart(\Xi_n)} &= \state¬executing \\
+    \remaining¬execunits^{(n + 1)} &= \remaining¬execunits^{(n)} - \rob¬execunits^{(n)}_{\find¬readytostart(\Xi_n)} \\
```
Yes, but it's not saturated per-se (because this will not trigger if there are not enough execution units available).

koute commented on text/pvm_invocations.tex:
```diff
@@ -292,15 +302,50 @@ \subsection{General Functions}\label{sec:generalfunctions}
+    \using g &= 100 + \max(0, \registers_7 - a - c) \cdot 10 \\
+    \tup{\execst', \gascounter', \registers'_7, (\mathbf{u}'_\ram¬access)\interval{a}{\registers_7}} &\equiv \begin{cases}
+      % Not enough gas to even cover the constant cost of the call, regardless of whether it's a no-op or not.
+      \tup{\panic, \gascounter, \registers_7, (\mathbf{u}_\ram¬access)\interval{a}{\registers_7}} &\when \gascounter < 100 \\
```
I did this mostly for clarity, taking advantage of the usual "if it's not explicitly defined for this specific case then the default happens" thing we have across the whole GP.

This hostcall is a little special in that if there's not enough gas available to execute it fully but enough gas to pay for the O(1) cost then it doesn't abort the execution, but essentially returns an "allocation failed" error to the user so that they can handle it gracefully if they want. (And arguably we maybe might want to make other expensive variable-cost hostcalls also work this way?). Basically, I only wanted to have a *single* `if-otherwise` block for the whole hostcall, instead of two (one just for `g` and another for the rest of the state), so it made sense to also roll the gas into that single block.

I suppose what I could do is this:

<img width="1140" height="344" alt="grow_heap" src="https://github.com/user-attachments/assets/4040f4b3-a9c1-4ae9-a18f-22ad41c902af" />

So then the gas subtraction and out-of-gas check will happen "outside" of the hostcall. But, this seems a little rube-goldberg-y to me, since we're already checking the gas inside of the hostcall anyway? What do you think?

koute committed:
Review comment: replace `PVM` with `\textsc{pvm}`

koute committed:
Review comment: fix type of Psi_H

koute committed:
Review comment: remove empty equation in Psi_H

koute committed:
Review comment: replace the symbol for the gas model simulation equation

koute committed:
Review comment: reword block-based gas charging sentence

koute committed:
Change `\panic` -> `\oog` for `grow_heap` hostcall

Ultimately this doesn't matter since panic and out-of-gas are equivalent
for toplevel VMs, but for consistency let's use \oog here.

0xjunha commented on text/pvm_invocations.tex:
```diff
@@ -292,15 +302,50 @@ \subsection{General Functions}\label{sec:generalfunctions}
+    \using c &= |\Nclamp{a}{b} \subseteq \writable{\memory}| \\
+    % The total gas cost, based on the number of pages that will become writable after this call.
+    \using g &= 100 + \max(0, \registers_7 - a - c) \cdot 10 \\
+    \tup{\execst', \gascounter', \registers'_7, (\mathbf{u}'_\ram¬access)\interval{a}{\registers_7}} &\equiv \begin{cases}
```
It seems \mathbf{u} is not well defined here, should be μ instead?

0xjunha reviewed (commented):
Left some comments on the new hostcall.

0xjunha reviewed (commented):
Probably a typo - for the recursion to carry gas-charged flag, `\Psi^*` looks correct here

0xjunha commented on text/pvm.tex:
```diff
@@ -714,31 +733,32 @@ \subsection{Host Call Definition}
-      \tup{\varepsilon', \imath', \gascounter', \registers', \mem', \mathbf{x}} &\when \varepsilon' \in \set{ \halt, \panic, \oog } \cup \set{\fault} \times \pvmreg \\[4pt]
+    \tuple{\set{\panic, \oog, \halt} \cup \set{\fault} \times \pvmreg, \pvmreg, \signedgas, \bool, \regs, \ram, X}\\
+    \tup{\pvm¬blob, \imath, \gascounter, \gaschargedflag, \registers, \mem, f, \mathbf{x}} &\mapsto \begin{cases}
+      \multicolumn{2}{l}{\text{let }(\varepsilon', \imath', \gascounter', \gaschargedflag', \registers', \mem') = \Psi(\pvm¬blob, \imath, \gascounter, \gaschargedflag, \registers, \mem):} \\[8pt]
```
Yes, it isn't recursion into `ΨH`. It's about calling the basic PVM with the flag: `Ψ` takes 5 params (without gas-charged flag) and `Ψ*` takes 6 (with the flag). Shouldn't we call `Ψ*` here since we want to carry the flag for per-block metering? (And also adjust (A.1) to use `Ψ*` for "continue" case)

koute committed:
Review comment: remove accidental extra 'the'

koute committed:
Remove \Psi^*; add gas-charged flag plumbing for inner PVMs too

koute committed:
Review comment: fixes to the `grow_heap` hostcall

koute committed:
Make it more clear what happens if a hostcall does its own gas metering

koute commented:
@0xjunha Thank you for the review!

zdave-parity commented on text/pvm.tex:
```diff
@@ -839,3 +842,414 @@ \subsection{Argument Invocation Definition}
+
+\begin{equation}
+  \Xi_0(\imath) = \abracegroup{
+    \imath^{(0)} &= \imath \\
```
The notation here is a bit confusing. What's the purpose of the superscripts on the state variables? They seem to just duplicate the overall state `\Xi` subscript. Is it so the individual variables at a particular step can be referenced? Why not just write eg `\Xi_n(i)_c` instead of `c^(n)`? This would be more consistent with the notation used elsewhere and defined in the notation section.

In any case it would be helpful here to define the type of the state `\Xi_n` and explain the meaning of each of the fields, as is done for most of the other tuples in the GP.

koute replied:
> Yes. I don't think we should invent new notation just for this section, we should stick to what is already defined in section 3.

Okay, fair enough. In my defense, I was operating under the assumption that I should maximize the readability (or at least what I *think* maximizes readability, which, to be fair, is subjective so not everyone might agree) and that a very slight deviation for this section (which is otherwise somewhat complex and self-contained) would be okay.

I'll rewrite it to match the existing conventions better.

zdave-parity commented on text/pvm.tex:
```diff
@@ -839,3 +842,414 @@ \subsection{Argument Invocation Definition}
+    \rob¬cyclesleft^{(0)} = \emptyorderedmap,
+    \rob¬deps^{(0)} = \emptyorderedmap,
+    \rob¬regs^{(0)} = \emptyorderedmap,
+    \rob¬execunits^{(0)} = \emptyorderedmap \\
```
It would be helpful here to define the type of the state `\Xi_n` and explain the meaning of each of the fields, as is done for most of the other tuples in the GP.

zdave-parity commented on text/pvm.tex:
```diff
@@ -839,3 +842,414 @@ \subsection{Argument Invocation Definition}
+    \rob¬deps^{(0)} = \emptyorderedmap,
+    \rob¬regs^{(0)} = \emptyorderedmap,
+    \rob¬execunits^{(0)} = \emptyorderedmap \\
+    \remaining¬execunits^{(0)} &= \{
```
This should use tuple notation: `\tup{\is{\ALU}{4}, \is{\LOAD}{4}, ...}`. Same applies to the overall definition of `\Xi`.

zdave-parity commented on text/pvm.tex:
```diff
@@ -839,3 +842,414 @@ \subsection{Argument Invocation Definition}
+
+\begin{equation}
+  \Xi_{n + 1}(\Xi_n, \mathbf{c}, \mathbf{k}) = \begin{cases}
+    \Xi_{n + 1}'(\Xi_n, \mathbf{c}, \mathbf{k}) &\when n = 0 \lor (\imath^{(n)} \neq \none \land \cost¬decodeslots(\mathbf{c}, \mathbf{k}, \imath^{(n)}) \leq \remaining¬decodeslots^{(n)} \land |\rob¬state^{(n)}| < 32) \\
```
The `n = 0 \lor` seems unnecessary here?

koute commented on text/pvm_invocations.tex:
```diff
@@ -267,6 +270,13 @@ \subsection{Accumulate Invocation}\label{sec:accumulateinvocation}
+\newcommand{\next¬hostcall}{
+  \thehostcall¬counter
+  \stepcounter{hostcall¬counter}
+}
```
Well, if we accidentally change the indices and no one notices in a review then I think we have bigger problems... (:

Nevertheless, I don't have a strong opinion about this and I'd be happy to revert this, but I'd like to get a consensus (or Gav's opinion) as to whichever we prefer (since e.g. AFAIK Davide liked this change, so... :P)

davxy replied:
I do not have a strong opinion either; But I found this TeX mastery fascinating :-)

zdave-parity commented on text/pvm.tex:
```diff
@@ -839,3 +842,414 @@ \subsection{Argument Invocation Definition}
+\newcommand*{\MUL}{\text{M}}
+\newcommand*{\DIV}{\text{D}}
+\newcommand*{\find¬readytostart}{\mathfrak{S}}
+\newcommand*{\emptyorderedmap}{\langle\,\rangle}
```
If you're going to define a new notation you should add it to the notation section. I expect no new notation is needed here though, you can probably use either sequences or dictionaries for the reorder buffer?

zdave-parity commented on text/pvm.tex:
```diff
@@ -839,3 +842,414 @@ \subsection{Argument Invocation Definition}
+    \ctr¬instructions^{(0)} = 0,
+    \remaining¬decodeslots^{(0)} = \const¬maxdecodepercycle,
+    \remaining¬startpercycle^{(0)} = \const¬maxstartpercycle \\
+    \rob¬state^{(0)} &= \emptyorderedmap,
```
It might be clearer to make the reorder buffer a sequence-of-tuples rather than a tuple-of-sequences? Seems like ndot is not really necessary; you could just use the length of this sequence?

zdave-parity commented on text/pvm.tex:
```diff
@@ -839,3 +842,414 @@ \subsection{Argument Invocation Definition}
+The state transition function of the simulation from step $n$ to step $n + 1$ is defined as follows:
+
+\begin{equation}
+  \Xi_{n + 1}(\Xi_n, \mathbf{c}, \mathbf{k}) = \begin{cases}
```
If `\mathbf{c}` and `\mathbf{k}` are clearly defined earlier and used as free variables throughout the PVM section then it's probably fine to just use them here. In that case though I would avoid binding them in any functions you define. If you do that you're essentially shadowing them? And I would maybe still add a bit of text to clarify that they continue to have the same meanings as earlier.

koute commented on text/pvm.tex:
```diff
@@ -839,3 +842,414 @@ \subsection{Argument Invocation Definition}
+  \Xi_{n + 1}''(\Xi_n) = \abracegroup{
+    \rob¬state^{(n + 1)}_{\find¬readytostart(\Xi_n)} &= \state¬executing \\
+    \remaining¬execunits^{(n + 1)} &= \remaining¬execunits^{(n)} - \rob¬execunits^{(n)}_{\find¬readytostart(\Xi_n)} \\
+    \remaining¬startpercycle^{(n + 1)} &= \remaining¬startpercycle^{(n)} - 1
```
Yes, this is correct. The idea is to handle that through decode costs and execution unit requirements.

This counter is mostly to cap the number of times (per cycle) we have to check the reorder buffer for the next instruction to start executing. Currently I've set `const¬maxstartpercycle` to 5, but this is somewhat arbitrary - there could actually be more than 5 instructions waiting to start in the ROB (assuming their resource requirements don't overlap), however such situation is rare enough in practice that it doesn't make sense to waste the compute (i.e. to slow down the gas cost model simulation and regress our compile times) on checking this (very heavy diminishing returns; I measured, and picked the 5 based on those measurements).

koute committed:
Rewrite gas cost model equations; align to review comments

koute committed:
Review comment: restore explicit hostcall numbers

koute commented:
@zdave-parity Thanks for the thorough review! I *think* I fixed/improved everything you pointed out? I've rewritten the equations to be more in the style of the rest of the GP, while putting a lot of effort into keeping things as readable as possible, and also made things more explicit where possible. Hopefully it looks much better now? Please do let me know if there's still anything you'd like me to change in this new version.

zdave-parity commented on text/pvm.tex:
```diff
@@ -839,3 +860,533 @@ \subsection{Argument Invocation Definition}
+
+\begin{equation}
+  \gas¬sim¬ty \equiv \tuple{
+    \isa{\imath}{\set{\pvmreg, \none}},
```
Think you want `\pvmreg \cup \set{\none}` here.

zdave-parity commented on text/pvm.tex:
```diff
@@ -839,3 +860,533 @@ \subsection{Argument Invocation Definition}
+\begin{equation}
+\begin{aligned}
+  \forall a,b \in \execunits¬ty;\forall \oplus \in \set{+, -}&: a \oplus b \equiv c\\
+  \where \forall k \in \set{\ALU, \LOAD, \STORE, \MUL, \DIV}&:c_k \equiv a_k \oplus b_k\\
```
I guess there should really be a `c \in \execunits¬ty` somewhere here.

zdave-parity commented on text/pvm.tex:
```diff
@@ -839,3 +860,533 @@ \subsection{Argument Invocation Definition}
+\end{aligned}
+\end{equation}
+
+We define function $\inst¬srcregs(\mathbf{c}, \mathbf{k}, \imath)$ which returns a set of source registers read by a given instruction at $\imath$, and the function $\inst¬dstregs(\mathbf{c}, \mathbf{k}, \imath)$ which returns a set of destination registers which are written by a given instruction, as described by the equations in \ref{sec:instructiontables}. This is regardless of whether those r ...
```
```suggestion
We define the function $\inst¬srcregs(\mathbf{c}, \mathbf{k}, \imath)$ which returns the set of source registers read by the instruction at $\imath$, and the function $\inst¬dstregs(\mathbf{c}, \mathbf{k}, \imath)$ which returns the set of destination registers which are written by a given instruction, as described by the equations in \ref{sec:instructiontables}. This is regardless of whether those registers would actually have been modified by that instruction when executed at runtime. \token{ecalli} is assumed to neither read nor write to any registers in this model.
```

zdave-parity commented on text/pvm.tex:
```diff
@@ -839,3 +860,533 @@ \subsection{Argument Invocation Definition}
+
+We define function $\inst¬srcregs(\mathbf{c}, \mathbf{k}, \imath)$ which returns a set of source registers read by a given instruction at $\imath$, and the function $\inst¬dstregs(\mathbf{c}, \mathbf{k}, \imath)$ which returns a set of destination registers which are written by a given instruction, as described by the equations in \ref{sec:instructiontables}. This is regardless of whether those r ...
+
+We also define function $\cost¬cycles(\mathbf{c}, \mathbf{k}, \imath)$ which returns the number of cycles a given instruction needs to finish execution, $\cost¬decodeslots(\mathbf{c}, \mathbf{k}, \imath)$ which returns the number of decoding slots necessary to decode it, and $\cost¬execunits(\mathbf{c}, \mathbf{k}, \imath)$ which returns the number of virtual CPU execution units required to start ...
```
We also define the function

zdave-parity commented on text/pvm.tex:
```diff
@@ -839,3 +860,533 @@ \subsection{Argument Invocation Definition}
+    \tup{\mathbf{c}, \mathbf{k}, \imath} &\mapsto \max(\gas¬sim¬state^{\text{final}}_{\ctr¬cycles} - 3, 1)\\
+  }{
+    \where &\gas¬sim¬state^{\text{final}} &=&\text{ }\gas¬sim(\mathbf{c}, \mathbf{k}, \gas¬sim¬state^{\text{init}})\\
+      &\gas¬sim¬state^{\text{init}} &=&\text{ }\tup{\imath, 0, \const¬maxdecodepercycle, \const¬maxstartpercycle, \tup{4, 4, 4, 1, 1}, \sequence{}}
```
```suggestion
      &\gas¬sim¬state^{\text{init}} &=&\text{ }\tup{\imath, 0, \const¬maxdecodepercycle, \const¬maxstartpercycle, \tup{4, 4, 4, 1, 1}, \sq{}}
```

zdave-parity commented on text/pvm.tex:
```diff
@@ -839,3 +860,533 @@ \subsection{Argument Invocation Definition}
+
+The gas cost model for the \textsc{pvm} is a simplified model of a modern CPU microarchitecture, heavily inspired by what's used by production-grade compilers to predict how much time a given piece of code will take. For each basic block in the program the model simulates its execution flow and computes the required number of virtual CPU cycles that would be needed to execute it.
+
+The gas simulation state type $\gas¬sim¬ty$ contains the current instruction counter $\imath$, a cycle counter $\ctr¬cycles$, the number of decode slots $\remaining¬decodeslots$ remaining in the current cycle, the maximum number of instructions $\remaining¬startpercycle$ which are still allowed to start execution in the current cycle, the remaining available execution units $\remaining¬execunits$ ...
```
```suggestion
A gas simulation state, of set $\gas¬sim¬ty$, contains the current instruction counter $\imath$, a cycle counter $\ctr¬cycles$, the number of decode slots $\remaining¬decodeslots$ remaining in the current cycle, the maximum number of instructions $\remaining¬startpercycle$ which are still allowed to start execution in the current cycle, the remaining available execution units $\remaining¬execunits$, and the reorder buffer $\rob$. Formally:
```

And similarly elsewhere; the GP doesn't generally refer to these sets as types.

koute replied:
I was basing this off of this snippet:

```
The inner \textsc{pvm} invocation host-calls, meanwhile, depend on an integrated \textsc{pvm} type, which we shall denote $\pvmguest$. It holds a \textsc{pvm} program blob, instruction counter and \textsc{ram}:
\begin{equation}\label{eq:pvmguest}
  \pvmguest \equiv \tuple{\isa{\pg¬code}{\blob}, \isa{\pg¬ram}{\ram}, \isa{\pg¬pc}{\pvmreg}}
\end{equation}
```

which *does* seem to explicitly call such a type tuple a "type" and enumerates what it contains? Do you want me to change that one too?

zdave-parity commented on text/pvm.tex:
```diff
@@ -839,3 +860,533 @@ \subsection{Argument Invocation Definition}
+      \gas¬sim¬state &\otherwhen \gas¬sim¬state_{\imath} = \none \land l = 0 \\
+      \gas¬sim(\mathbf{c}, \mathbf{k}, \gas¬sim'''(\gas¬sim¬state)) &\otherwise \\
+    \end{cases}\\
+    \where l &= |\set{\build{i}{ i \in \N, \in¬rob¬field{i}{\rob¬state} \neq \none }}|
```
Not sure it's really defined what `s[i]` means when `i >= |s|`. Here you can avoid this by writing eg:

    \len{\sq{\build{r}{ r \orderedin \in¬rob, r_{\rob¬state} \neq \none }}}

zdave-parity commented on text/pvm.tex:
```diff
@@ -839,3 +860,533 @@ \subsection{Argument Invocation Definition}
+          \in¬rob¬field{j}{\rob¬regs} \setminus \inst¬dstregs(\mathbf{c}, \mathbf{k}, \gas¬sim¬state_{\imath}) &\otherwise
+        \end{cases} \\
+      \end{aligned} \\
+      j &\in \N
```
I think you want to write this like:

    forall j in N_|x_r| : x'_r[j]_r = ...

So quantify j properly and limit to just the valid indices.

zdave-parity commented on text/pvm.tex:
```diff
@@ -839,3 +860,533 @@ \subsection{Argument Invocation Definition}
+    \tup{\blob, \bitstring, \gas¬sim¬ty} &\to \gas¬sim¬ty \\
+    \tup{\mathbf{c}, \mathbf{k}, \gas¬sim¬state} &\mapsto \gas¬sim¬state'
+  }{
+    \where \gas¬sim¬state' &= \gas¬sim¬state\text{ except the following:}\\
```
Don't think you need "the following", the rest of the GP just uses "except:" and I think that reads more clearly.

zdave-parity commented on text/pvm.tex:
```diff
@@ -839,3 +860,533 @@ \subsection{Argument Invocation Definition}
+    \gas¬sim¬state'_{\remaining¬decodeslots} &= \remaining¬decodeslots - \cost¬decodeslots(\mathbf{c}, \mathbf{k}, \gas¬sim¬state_{\imath}) \\
+    \rob¬entry{\rob'}{j} &= \in¬rob¬entry{j}\text{ except: } \rob¬entry¬field{\rob'}{j}{\rob¬regs} = \in¬rob¬field{j}{\rob¬regs} \setminus \rob¬regs'\\
+    j &\in \N \\
+    \rob¬regs' &= \inst¬dstregs(\mathbf{c}, \mathbf{k}, \gas¬sim¬state_{\imath}) \\
```
Can probably remove the 's from these `rob` variables here.

zdave-parity commented on text/pvm.tex:
```diff
@@ -839,3 +860,533 @@ \subsection{Argument Invocation Definition}
+      \gas¬sim¬state_{\imath} + 1 + \Fskip(\gas¬sim¬state_{\imath}) &\otherwise
+    \end{cases}\\
+    \gas¬sim¬state'_{\remaining¬decodeslots} &= \remaining¬decodeslots - \cost¬decodeslots(\mathbf{c}, \mathbf{k}, \gas¬sim¬state_{\imath}) \\
+    \rob¬entry{\rob'}{j} &= \in¬rob¬entry{j}\text{ except: } \rob¬entry¬field{\rob'}{j}{\rob¬regs} = \in¬rob¬field{j}{\rob¬regs} \setminus \rob¬regs'\\
```
Should write this like:

    r' = x_r except: forall j in N_|x_r|: r'[j]_r = ...

zdave-parity commented on text/pvm.tex:
```diff
@@ -839,3 +860,533 @@ \subsection{Argument Invocation Definition}
+    \rob¬regs' &= \inst¬dstregs(\mathbf{c}, \mathbf{k}, \gas¬sim¬state_{\imath}) \\
+    \rob¬cyclesleft' &= \cost¬cycles(\mathbf{c}, \mathbf{k}, \gas¬sim¬state_{\imath}) \\
+    \rob¬execunits' &= \cost¬execunits(\mathbf{c}, \mathbf{k}, \gas¬sim¬state_{\imath}) \\
+    \rob¬deps' &= \set{\build{i}{ i \in N, \inst¬srcregs(\mathbf{c}, \mathbf{k}, \gas¬sim¬state_{\imath}) \cap \in¬rob¬field{i}{\rob¬regs} \neq \emptyset }} \\
```
```suggestion
    \rob¬deps' &= \set{\build{i}{ i \in \N_{\len{\in¬rob}}, \inst¬srcregs(\mathbf{c}, \mathbf{k}, \gas¬sim¬state_{\imath}) \cap \in¬rob¬field{i}{\rob¬regs} \neq \emptyset }} \\
```

zdave-parity commented on text/pvm.tex:
```diff
@@ -839,3 +860,533 @@ \subsection{Argument Invocation Definition}
+    \rob¬cyclesleft' &= \cost¬cycles(\mathbf{c}, \mathbf{k}, \gas¬sim¬state_{\imath}) \\
+    \rob¬execunits' &= \cost¬execunits(\mathbf{c}, \mathbf{k}, \gas¬sim¬state_{\imath}) \\
+    \rob¬deps' &= \set{\build{i}{ i \in N, \inst¬srcregs(\mathbf{c}, \mathbf{k}, \gas¬sim¬state_{\imath}) \cap \in¬rob¬field{i}{\rob¬regs} \neq \emptyset }} \\
+    \gas¬sim¬state'_{\rob} &= \rob' \concat \tuple{\state¬decoding, \rob¬cyclesleft', \rob¬deps', \rob¬regs', \rob¬execunits'} \\
```
```suggestion
    \gas¬sim¬state'_{\rob} &= \rob' \append \tuple{\state¬decoding, \rob¬cyclesleft', \rob¬deps', \rob¬regs', \rob¬execunits'} \\
```

zdave-parity commented on text/pvm.tex:
```diff
@@ -839,3 +860,533 @@ \subsection{Argument Invocation Definition}
+
+\begin{equation}
+  \find¬readytostart\colon \abracegroup{
+    \gas¬sim¬ty &\to \set{\N, \none}\\
```
```suggestion
    \gas¬sim¬ty &\to \N \cup \set{\none}\\
```

zdave-parity commented on text/pvm.tex:
```diff
@@ -839,3 +860,533 @@ \subsection{Argument Invocation Definition}
+    \gas¬sim¬state'_{\remaining¬startpercycle} &= \gas¬sim¬state_{\remaining¬startpercycle} - 1\\
+    n &= \find¬readytostart(\gas¬sim¬state)\\
+    \gas¬sim¬state'_{\remaining¬execunits} &= \gas¬sim¬state_{\remaining¬execunits} - \in¬rob¬field{n}{\rob¬execunits}\\
+    \gas¬sim¬state'_{\rob} &= \in¬rob\text{ except: } \rob¬entry¬field{\rob'}{n}{\rob¬state} = \state¬executing
```
```suggestion
    \gas¬sim¬state'_{\rob} &= \in¬rob\text{ except: } \rob¬entry¬field{\gas¬sim¬state'_{\rob}}{n}{\rob¬state} = \state¬executing
```

zdave-parity commented on text/pvm.tex:
```diff
@@ -839,3 +860,533 @@ \subsection{Argument Invocation Definition}
+\begin{equation}
+  \find¬readytostart\colon \abracegroup{
+    \gas¬sim¬ty &\to \set{\N, \none}\\
+    \gas¬sim¬state &\mapsto \min(j \in \N : \in¬rob¬field{j}{\rob¬state} = \state¬waiting \land \in¬rob¬field{j}{\rob¬execunits} \leq \gas¬sim¬state_{\remaining¬execunits} \land (\forall k \in \in¬rob¬field{j}{\rob¬deps} \Rightarrow \in¬rob¬field{k}{\rob¬cyclesleft} \leq 0))
```
I guess this is assuming `\min(\none) = \none`, not sure that is really defined anywhere though. I would probably just be explicit about this here for clarity, maybe something like:

    x -> \none if r = \none, min(r) otherwise
    where r = { j in \N_{\len{x_r}} | ... }

Separately, I don't think `\Rightarrow` should be used here. That is used to mean "implies" elsewhere, think you just want a colon here?

zdave-parity commented on text/pvm.tex:
```diff
@@ -839,3 +860,533 @@ \subsection{Argument Invocation Definition}
+    \rob¬entry{\gas¬sim¬state'_{\rob}}{j} &= \in¬rob¬entry{j}\text{ except the following:} \\
+    &\begin{aligned}
+      \rob¬entry¬field{\gas¬sim¬state'_{\rob}}{j}{\rob¬state} &= \begin{cases}
+        \none &\when \forall k \in \N,\quad 0 \leq k \leq j \Rightarrow \in¬rob¬field{k}{\rob¬state} = \state¬executed \\
```
```suggestion
        \none &\when \forall k \in \N_{j+1},\quad \in¬rob¬field{k}{\rob¬state} \in \set{\state¬executed, \none} \\
```

zdave-parity commented on text/pvm.tex:
```diff
@@ -839,3 +860,533 @@ \subsection{Argument Invocation Definition}
+      \rob¬entry¬field{\gas¬sim¬state'_{\rob}}{j}{\rob¬state} &= \begin{cases}
+        \none &\when \forall k \in \N,\quad 0 \leq k \leq j \Rightarrow \in¬rob¬field{k}{\rob¬state} = \state¬executed \\
+        \state¬waiting &\when \in¬rob¬field{j}{\rob¬state} = \state¬decoding \\
+        \state¬executed &\when \in¬rob¬field{j}{\rob¬state} = \state¬executing \land \in¬rob¬field{j}{\rob¬cyclesleft} = 0 \\
```
Is it intended that we stay in the EXE state for 1 extra cycle? Doesn't seem like this changes much other than how long it takes for buffer slots to free up, but it does seem a bit weird. In particular cycles will go negative but it's supposed to be `\in \N`.

koute replied:
> Is it intended that we stay in the EXE state for 1 extra cycle?

Yes, this is intended as what matters are the transitions when the cycle counter ticks over.

Let's say we have an instruction with a cycle count of 1 cycle; triggering the state machine we get:

- Step 0: cycle 0 -> 0: instruction gets decoded, state is now DEC
- Step 1: cycle 0 -> 1: instruction is marked as waiting, state is transitioned DEC -> WAIT
- Step 2: cycle 1 -> 1: instruction is marked for execution, state is transitioned WAIT -> EXE
- Step 3: cycle 1 -> 2: cycle count is decremented and becomes 0, state unchanged EXE -> EXE
- Step 4: cycle 2 -> 3: cycle count is decremented and becomes -1, state is transitioned EXE -> FIN
- Step 5: cycle 3 -> 4: state is now FIN -> retired

Or if we only look at the cycle transitions start-to-end:

- Cycle 0 -> 1: state went DEC -> EXE
- Cycle 1 -> 2: state went EXE -> EXE
- Cycle 2 -> 3: state went EXE -> FIN
- Cycle 3 -> 4: state went FIN -> retired

So, it gets decoded, then executed, then marked as finished, then it's retired. What matters is for how many cycle *transitions* it *stays* as EXE (which is only one here, which I suppose intuitively makes sense because CPUs do "work" only when the clock ticks).

We can also put this in a timeline graph form, assigning each type of transition a letter:

```
0123
DeER
```

And here's what LLVM's microarchitectural simulator prints out if I ask it to simulate a 1-cycle instruction like `add`:

```
[0,0]     DeER .    . .   add	rax, rbx
```

...which is not a coincidence. (: 

> In particular cycles will go negative but it's supposed to be `\in \N`.

Ah yeah, good point. Practically this doesn't matter, so I suppose I can just clamp them?

zdave-parity replied:
> Practically this doesn't matter, so I suppose I can just clamp them?

Sure.

zdave-parity commented on text/pvm.tex:
```diff
@@ -839,3 +860,533 @@ \subsection{Argument Invocation Definition}
+        \in¬rob¬field{j}{\rob¬cyclesleft} - 1 &\when \in¬rob¬field{j}{\rob¬state} = \state¬executing \\
+        \in¬rob¬field{j}{\rob¬cyclesleft} &\otherwise \\
+      \end{cases} \\
+      \rob¬entry¬field{\gas¬sim¬state'_{\rob}}{j}{\rob¬regs} &= \begin{cases}
```
This bit doesn't seem necessary, dependencies that have no cycles left have no effect?

koute replied:
Hmm... ah yep, you're right. Good catch! Yeah, this is here because I originally wrote these equations based on my code where I *do* need to do this part, but after simplifying everything for the GP this is now not necessary.

zdave-parity commented on text/pvm.tex:
```diff
@@ -839,3 +860,533 @@ \subsection{Argument Invocation Definition}
+  }{
+    \where \gas¬sim¬state' &= \gas¬sim¬state\text{ except the following:}\\
+    \gas¬sim¬state'_{\remaining¬decodeslots} &= \const¬maxdecodepercycle, \gas¬sim¬state'_{\remaining¬startpercycle} = \const¬maxstartpercycle, \gas¬sim¬state'_{\ctr¬cycles} = \gas¬sim¬state_{\ctr¬cycles} + 1\\
+    \gas¬sim¬state'_{\remaining¬execunits} &= \gas¬sim¬state_{\remaining¬execunits} + \sum_{n \in \N \Rightarrow \in¬rob¬field{n}{\rob¬state} = \state¬executing \land \in¬rob¬field{n}{\rob¬cyclesleft} = 1} \in¬rob¬field{j}{\rob¬execunits} \\
```
```suggestion
    \gas¬sim¬state'_{\remaining¬execunits} &= \gas¬sim¬state_{\remaining¬execunits} + \sum_{n \in \N_{\len{\in¬rob}}, \in¬rob¬field{n}{\rob¬state} = \state¬executing \land \in¬rob¬field{n}{\rob¬cyclesleft} = 1} \in¬rob¬field{n}{\rob¬execunits} \\
```

zdave-parity commented on text/pvm.tex:
```diff
@@ -839,3 +860,533 @@ \subsection{Argument Invocation Definition}
+    \where \gas¬sim¬state' &= \gas¬sim¬state\text{ except the following:}\\
+    \gas¬sim¬state'_{\remaining¬decodeslots} &= \const¬maxdecodepercycle, \gas¬sim¬state'_{\remaining¬startpercycle} = \const¬maxstartpercycle, \gas¬sim¬state'_{\ctr¬cycles} = \gas¬sim¬state_{\ctr¬cycles} + 1\\
+    \gas¬sim¬state'_{\remaining¬execunits} &= \gas¬sim¬state_{\remaining¬execunits} + \sum_{n \in \N \Rightarrow \in¬rob¬field{n}{\rob¬state} = \state¬executing \land \in¬rob¬field{n}{\rob¬cyclesleft} = 1} \in¬rob¬field{j}{\rob¬execunits} \\
+    \rob¬entry{\gas¬sim¬state'_{\rob}}{j} &= \in¬rob¬entry{j}\text{ except the following:} \\
```
```suggestion
    \gas¬sim¬state'_{\rob} &= \in¬rob\text{ except } \forall j \in \N_{\len{\in¬rob}}: \\
```

(And drop the `j \in \N` at the bottom)

zdave-parity commented on text/pvm.tex:
```diff
@@ -839,3 +860,533 @@ \subsection{Argument Invocation Definition}
+  \mathfrak{m} \equiv 25
+\end{equation}
+
+The cost of a branch depends on whether any of its targets (either the jump target or the implicit fallthrough) points to an instruction byte which is equal to the opcode for the \token{unlikely} or the \token{trap} instruction; formally:
```
```suggestion
The cost of a branch depends on whether any of its targets (either the jump target or the implicit fallthrough) point to an instruction byte which is equal to the opcode for the \token{unlikely} or the \token{trap} instruction; formally:
```

zdave-parity reviewed (commented):
Thanks, the gas cost model section looks much better to me now. Just a few nits there.

zdave-parity commented on text/pvm.tex:
```diff
@@ -33,26 +30,28 @@ \subsection{Basic Definition}
-      \tup{\oog, \imath, \gascounter', \registers, \mem} &\when \gascounter' < 0\\
-      \tup{\varepsilon, 0, \gascounter', \registers', \mem'} &\when \varepsilon \in \set{ \panic, \halt }\\
-      \tup{\varepsilon, \imath, \gascounter', \registers, \mem} &\otherwise
+    \tuple{\blob, \pvmreg, \gas, \bool, \regs, \ram} &\to \tuple{\set{\halt, \panic, \oog} \cup \set{\fault, \host} \times \pvmreg, \pvmreg, \signedgas, \bool, \regs, \ram}\\
```
Think we don't need `\signedgas` here any more? As the behaviour now is that gas is not consumed by regular execution if it would drop below 0. We still need signed gas higher up for host calls.

zdave-parity commented on text/pvm.tex:
```diff
@@ -33,26 +30,28 @@ \subsection{Basic Definition}
+    \tuple{\blob, \pvmreg, \gas, \bool, \regs, \ram} &\to \tuple{\set{\halt, \panic, \oog} \cup \set{\fault, \host} \times \pvmreg, \pvmreg, \signedgas, \bool, \regs, \ram}\\
+    \tup{\pvm¬blob, \imath, \gascounter, \gaschargedflag, \registers, \mem} &\mapsto \begin{cases}
+      \Psi(\pvm¬blob, \imath', \gascounter', \gaschargedflag', \registers', \mem') &\when \varepsilon = \continue\\
+      \tup{\oog, \imath, \gascounter', \gaschargedflag', \registers, \mem} &\when \gascounter' < 0\\
```
Think this case can just be removed now? `\gascounter'` is never <0 AFAICT but OOG will be handled fine by the last case here?

zdave-parity commented on text/pvm.tex:
```diff
@@ -124,35 +123,62 @@ \subsection{Basic Blocks and Termination Instructions}
+For convenience we will define function $\startofbasicblock$ which represents the start of a basic block for any $\imath$ within that basic block. Formally:
+
+\begin{equation}
+  \startofbasicblock\colon \abracegroup{
```
I think there are a few cases here where things aren't well-defined. eg If we start executing at 0 but there isn't a valid instruction at 0 then the gas charging stuff doesn't seem well-defined -- there is no basic block at 0, so what is charged? I wonder if more generally we should require that the initial program counter is at the start of a basic block; I wouldn't be surprised if some implementations required this now, but it doesn't seem to be a requirement here.

koute replied:
> I wonder if more generally we should require that the initial program counter is at the start of a basic block

I'd love to be able to do that, unfortunately we can't require this as we also need the ability to resume in the middle of a basic block (e.g. for an inner VM which previously was interrupted due to a page fault).

That said, jumps *within* the program are guaranteed to be valid and can only go to the beginning of basic blocks, so it's mostly a question of what to do on the initial entry. Hmmm...

I think you're right in that some of the teams are most likely handling these corner cases differently.

I suppose maybe the most prudent thing to do right now would be to just prevalidate the program before we start execution, and reject it if it contains any abnormalities? That would be relatively simple to do and unambiguous. For JAM 1.0 at very least we'll be assuming that everyone's going to be running a whole program recompiler anyway, which implies that we'll be paying O(n) cost anyway per unique program invocation (no fancy recompiler/interpreter hybrids nor lazy JIT recompilation) which means that such prevalidation would effectively be "free". And we could always relax this in the future.

Thoughts?

zdave-parity replied:
> That said, jumps within the program are guaranteed to be valid and can only go to the beginning of basic blocks, so it's mostly a question of what to do on the initial entry. Hmmm...

Yes, I meant initial entry. Resumption seems easier as you only care about the "native" PC for that? At least I don't think there's any way to set or inspect the PC of a VM?

Prevalidation sounds reasonable to me. As you say it shouldn't really cost anything with a recompiler and would close off some attack surface.

koute replied:
> At least I don't think there's any way to set or inspect the PC of a VM?

Indeed. You can only set the program counter when you start the VM (with `invoke`), and you can only fetch it when you end it (with `expunge`). Otherwise it's inaccessible to keep things simple.

zdave-parity commented on text/pvm.tex:
Not sure this is really true any more?

zdave-parity commented on text/pvm.tex:
```diff
@@ -124,35 +123,62 @@ \subsection{Basic Blocks and Termination Instructions}
   \using \mathbf{x} &= \set{\build{x}{x \in \mathbf{r} \wedge x \bmod 2^{32} \not\in \readable\mem\ \vee\ x \in \mathbf{w} \wedge x \bmod 2^{32} \not\in \writable\mem}} \\
   \tup{\varepsilon^*, \imath^*, \gascounter^*, \registers^*, \mem^*} &= \begin{cases}
+    \tup{\oog, \imath, \gascounter', \registers, \mem} &\when \tilde{\varepsilon} = \oog \\
     \tup{\varepsilon, \imath', \gascounter', \registers', \mem'} &\when \mathbf{x} = \emset \\
```
Think you ought to change these `\when` to `\otherwhen` now?

zdave-parity commented on text/pvm.tex:
```diff
@@ -124,35 +123,62 @@ \subsection{Basic Blocks and Termination Instructions}
+
 During the course of executing instructions \textsc{ram} may be accessed. When an index of \textsc{ram} below $2^{16}$ is required, the machine always panics immediately without further changes to its state regardless of the apparent (in)accessibility of the value. Otherwise, should the given index of \textsc{ram} not be accessible then machine state remains unchanged and the exit reason is a fau ...
 
 Formally, let $\mathbf{r}$ and $\mathbf{w}$ be the set of indices by which $\mem$ must be subscripted for inspection and mutation respectively in order to calculate the result of $\Psi_1$. We define the memory-access exceptional execution state $\varepsilon^\mu$ which shall, if not $\continue$, singly effect the returned return of $\Psi_1$ as following:
```
This paragraph could do with rewording a bit I think. `\varepsilon^\mu` doesn't actually exist AFAICT and this equation now also covers out-of-gas.

Perhaps it would be neatest to actually define `\varepsilon^\mu` (as either `\continue`, `\panic`, or `\fault`) and then separately define the `*` variables based on `\varepsilon^\mu` and `\tilde{\varepsilon}`?

zdave-parity commented on text/pvm.tex:
```diff
@@ -124,35 +123,62 @@ \subsection{Basic Blocks and Termination Instructions}
+
+\begin{equation}
+  \gaschargedflag^* = \begin{cases}
+    \bot &\when \gaschargedflag' = \bot \lor \imath' < \imath \lor \startofbasicblock(\imath') \ne \startofbasicblock(\imath) \\
```
I think you want <= rather than < for the restarted-block condition? As presumably a jump/branch can jump to itself. I guess we want to avoid triggering this in the case of a fault, where we don't advance over the instruction. I _think_ this is fine as we're using `\imath'` rather than `\imath^*` here. It's all a bit subtle though and maybe there is a better way to formalise this...

koute replied:
Hmm... maybe it'd be better to delete this and have the control flow instructions explicitly reset the gas charged flag?

zdave-parity replied:
Sounds reasonable. Or maybe rather than detecting the PC changing, check if (a) the opcode is in the "ends basic block" set and (b) was actually executed?

koute committed:
More cleanups; align with most recent review comments

zdave-parity commented on text/pvm_invocations.tex:
```diff
@@ -273,7 +276,7 @@ \subsection{General Functions}\label{sec:generalfunctions}
 Unlike the Accumulate functions in appendix \ref{sec:accumulatefunctions}, these do not mutate an accumulation context. Some, such as $\mathtt{write}$ mutate a service account and both accept and return some $\mathbf{s} \in \serviceaccount$. Others are more general functions, such as $\mathtt{fetch}$ and do not assume any context but have a parameter list suffixed with an ellipsis to denote that  ...
 
-Other than the gas-counter which is explicitly defined, elements of \textsc{pvm} state are each assumed to remain unchanged by the host-call unless explicitly specified.
+Elements of \textsc{pvm} state are each assumed to remain unchanged by the host-call unless explicitly specified. If a given hostcall doesn't explicitly define $\gascounter'$ then the following holds:
```
This is not entirely accurate as we still want some of the "except as indicated below" stuff for `grow_heap`. It doesn't specify for example `\mathbf{s}'`.

zdave-parity commented on text/pvm_invocations.tex:
```diff
@@ -299,8 +302,43 @@ \subsection{General Functions}\label{sec:generalfunctions}
+      % Not enough gas to even cover the constant cost of the call, regardless of whether it's a no-op or not.
+      \tup{\oog, \gascounter, \registers_7, (\memory_\ram¬access)\interval{a}{\registers_7}} &\when \gascounter < 100 \\
+      % Expand the accessible RW data region, and return the new page count.
+      \tup{\continue, \gascounter - g, \max(a + c, \registers_7), \mathrm{W}} &\otherwhen \registers_7 \leq b \land \gascounter \ge g \\
```
`\mathrm{W}` is not quite right as the LHS is a sequence. I think it would be reasonable to write something like `[W, ..., W]`?

zdave-parity commented on text/pvm_invocations.tex:
```diff
@@ -299,8 +302,43 @@ \subsection{General Functions}\label{sec:generalfunctions}
+    \using g &= 100 + \max(0, \registers_7 - a - c) \cdot 10 \\
+    \tup{\execst', \gascounter', \registers'_7, (\memory'_\ram¬access)\interval{a}{\registers_7}} &\equiv \begin{cases}
+      % Not enough gas to even cover the constant cost of the call, regardless of whether it's a no-op or not.
+      \tup{\oog, \gascounter, \registers_7, (\memory_\ram¬access)\interval{a}{\registers_7}} &\when \gascounter < 100 \\
```
phi_7 may be < a, probably you want to define some intermediate variable equal to say max(a, phi_7) to use in these ranges? I'm not sure it's well defined what `x_{a..b}` means when b<a.

zdave-parity commented on text/pvm_invocations.tex:
```diff
@@ -299,8 +302,43 @@ \subsection{General Functions}\label{sec:generalfunctions}
+      %
+      % This will also trigger when the user doesn't have enough gas to request as many pages
+      % as they want, in which case they will be able to gracefully handle this situation.
+      \tup{\continue, \gascounter - 100, a + c, (\memory_\ram¬access)\interval{a}{\registers_7}} &\otherwise \\
```
It's a bit weird that we return `\continue` in the OOG case?

koute replied:
Well, generally the idea was that we could let the service handle it gracefully. Unlike e.g. accessing invalid memory running out of gas isn't necessarily abnormal, and *if* it's a costly hostcall it's conceivable that they might not have enough gas to execute this particular hostcall itself, but still have enough gas to recover? And if the desired behavior is "abort" then they could do just it themselves, instead of us forcing them. (Especially for a memory allocation function, where "not enough memory" is a reasonable enough error condition that people *do* handle in certain domains.)

That said, this was designed *before* I/we decided to go for per-page gas metering, which means that the final gas costs for this host-call will most likely *not* be very expensive (it won't *eagerly* allocate memory at very least). So I suppose the question is: do we want *in general* (as in - as a general design principle for *all*/most hostcalls) to force the OOG, or maybe it'd be nice to let the user decide what to do themselves? (In which case also tweaking other hostcalls like this could make sense.) Thoughts?

zdave-parity replied:
My preference would be to just always abort on OOG. I'm not sure you can sensibly handle OOG in the code that is running out of gas... what if you have just enough gas to execute the host call then run out of gas right after? You're going to abort and you can't do anything about it.

If you want to change behaviour based on remaining gas I think you need to do this up-front, by eg calling checkpoint (so you can optimistically try to do more, but with any changes automatically getting rolled back if there isn't enough gas) or inspecting the remaining gas and making a decision based on that.

koute replied:
Okay, I simplified it; now it will just abort with OOG, and also simplified the equations a little bit.

zdave-parity reviewed (commented):
Hopefully last batch of comments!

koute committed:
Prevalidate PVM blobs; immediately panic on bad instruction pointer

At very least for JAM 1.0 we will we be requiring an ahead-of-time
whole-program recompiler for PVM programs, so it makes sense to also
prevalidate the program blobs ahead of time, effectively removing a number
of corner cases that otherwise would have to be explicitly handled.
The expectation is that the prevalidation can be done essentially
"for free" during the recompilation process.

We also now immediately panic if a given instruction pointer is not
deemed valid.

koute committed:
Remove the paragraph about the gas always being charged

koute committed:
Clean up the single-step state transition equations

koute committed:
Reset the gas flag when we leave a basic block terminating instruction

koute committed:
Clean up `grow_heap` hostcall

koute committed:
`\Psi_1` and `\Psi` now don't return negative gas

koute committed:
Fix hostcall exception wording

koute commented:
@zdave-parity Okay, every comment should be addressed (I *think*), except the OOG case in `grow_heap` (which I'm happy to address whichever way you decide). As discussed, I also added the PVM blob prevalidation (validate the program blob structure wrt the basic blocks, make sure there are no invalid opcodes, and also insta-panic if the initial entry-point is invalid) which should get rid of a bunch of corner cases. (I added it to `deblob` since that is very conveniently used for both toplevel and inner PVM invocations; might be a good idea to rename it, but I don't really have any good idea for a better name.)

zdave-parity commented on text/pvm.tex:
```diff
@@ -839,3 +866,534 @@ \subsection{Argument Invocation Definition}
+  }\\
+\end{equation}
+
+A reorder buffer entry state, of set $\rob¬entry¬ty$, contains its current state $\rob¬state$, the number of cycles left $\rob¬cyclesleft$, a set of reorder buffer indices considered its dependencies $\rob¬deps$, a set of clobbered registers $\rob¬regs$, and the execution units used $\rob¬execunits$. Formally:
```
```suggestion
A reorder buffer entry, of set $\rob¬entry¬ty$, contains its current state $\rob¬state$, the number of cycles left $\rob¬cyclesleft$, a set of reorder buffer indices considered its dependencies $\rob¬deps$, a set of clobbered registers $\rob¬regs$, and the execution units used $\rob¬execunits$. Formally:
```

zdave-parity commented on text/pvm.tex:
```diff
@@ -839,3 +866,534 @@ \subsection{Argument Invocation Definition}
+  }\\
+\end{equation}
+
+A set $\execunits¬ty$ maps each execution unit kind (${\ALU, \LOAD, \STORE, \MUL, \DIV}$) into a number:
```
```suggestion
A tuple of the set $\execunits¬ty$ maps each execution unit kind (${\ALU, \LOAD, \STORE, \MUL, \DIV}$) into a number:
```

zdave-parity commented on text/pvm.tex:
```diff
@@ -839,3 +866,534 @@ \subsection{Argument Invocation Definition}
+  }
+\end{equation}
+
+For convenience we define addition and subtraction of two sets $\execunits¬ty$ to be equivalent to the memberwise operation of the same kind; formally:
```
```suggestion
For convenience we define addition and subtraction of two tuples of the set $\execunits¬ty$ to be equivalent to the memberwise operation of the same kind; formally:
```

zdave-parity commented on text/pvm.tex:
```diff
@@ -839,3 +866,534 @@ \subsection{Argument Invocation Definition}
+\end{aligned}
+\end{equation}
+
+Similarly, a comparison between two sets $\execunits¬ty$ is true if and only if the chosen comparison holds for all corresponding members:
```
```suggestion
Similarly, a comparison between two tuples of the set $\execunits¬ty$ is true if and only if the chosen comparison holds for all corresponding members:
```

zdave-parity commented on text/pvm.tex:
```diff
@@ -33,40 +30,50 @@ \subsection{Basic Definition}
+      \tuple{\blob, \bitstring, \pvmreg} &\to \bool \\
+      (\mathbf{c}, \mathbf{k}, \imath) &\mapsto \begin{cases}
+        \top &\when \imath > 0 \land \imath = \len{\mathbf{k}}\\
+        \mathfrak{v}(\mathbf{c}, \mathbf{k}, \imath + 1 + \text{skip}(\imath)) &\otherwhen \mathfrak{v}_{\text{loc}}(\mathbf{c}, \mathbf{k}, \imath)\\
```
```suggestion
        \mathfrak{v}_{\text{blob}}(\mathbf{c}, \mathbf{k}, \imath + 1 + \text{skip}(\imath)) &\otherwhen \mathfrak{v}_{\text{loc}}(\mathbf{c}, \mathbf{k}, \imath)\\
```

zdave-parity commented on text/pvm.tex:
```diff
@@ -124,34 +131,70 @@ \subsection{Basic Blocks and Termination Instructions}
 \end{align}
 
-We define $\varepsilon$ together with the posterior values of regular execution (denoted as prime) of each of the items of the machine state as being in accordance with the table below. When transitioning machine state for an instruction, a number of conditions typically hold true and instructions are defined essentially by their exceptions to these rules. Specifically, the machine does not halt, ...
+We also have to adjust the state based on the posterior value of the instruction counter, to force a gas charge on the next step when necessary:
```
```suggestion
We also have to adjust the state based on the executed instruction, to force a gas charge on the next step when necessary:
```

(Posterior value of instruction counter is no longer inspected)

zdave-parity commented on text/pvm.tex:
```diff
@@ -124,34 +131,70 @@ \subsection{Basic Blocks and Termination Instructions}
 \begin{equation}
-  \varepsilon = \continue,\quad \imath' = \imath + 1 + \Fskip(\imath),\quad \gascounter' = \gascounter - \gascost,\quad \registers' = \registers,\quad\mem' = \mem \text{ except as indicated }
+  \gaschargedflag^* = \begin{cases}
+    \bot &\when \gaschargedflag' = \bot \lor (\mathbf{c}_\imath \in T \land \varepsilon^* \in \set{\continue} \cup \set{\host} \times \pvmreg) \\
```
```suggestion
    \bot &\when \gaschargedflag' = \bot \lor (\instructions_\imath \in T \land \varepsilon^* \in \set{\continue} \cup \set{\host} \times \pvmreg) \\
```

Think you need `\instructions` here rather than `\mathbf{c}` as we can execute past the end? Possibly as part of validation you could require that the final instruction is a basic-block terminator, that would make this case go away.

zdave-parity commented on text/pvm.tex:
```diff
@@ -33,40 +30,50 @@ \subsection{Basic Definition}
+We declare the general \textsc{pvm} function $\Psi$. We assume a single-step invocation function define $\Psi_1$ and define the full \textsc{pvm} recursively as a sequence of such mutations up until the single-step mutation results in a halting condition. We additionally define the function $\deblob$ which extracts the instruction data, opcode bitmask and dynamic jump table from a \textsc{pvm} pr ...
 \begin{align}
   \Psi&\colon \abracegroup{
-    \tuple{\blob, \pvmreg, \gas, \regs, \ram} &\to \tuple{\set{\halt, \panic, \oog} \cup \set{\fault, \host} \times \pvmreg, \pvmreg, \signedgas, \regs, \ram}\\
```
The signature of `\Psi` is also given in the overview section, should replace `\signedgas` with `\gas` there. There is some text talking about negative gas there which is now not really true.

zdave-parity reviewed (dismissed):
Thanks. Few trivial things from the latest commits. Happy for this to be merged otherwise.

koute committed:
Wording cleanups

koute committed:
Enforce jump target validity; if *any* branch is invalid, panic

koute committed:
Enforce the last instruction to be a basic block terminator

koute committed:
Move `\abracegrouptwo` to preamble

koute committed:
Make `deblob` def. slightly nicer; prevent it from clipping the column

koute committed:
Clarify where `U`, `T` and `skip` come from in `deblob`

koute committed:
Adjust \Psi and the attached description in the overview section

koute commented:
I took the opportunity and added stricter validation of jumps/branches (e.g. for a branch if either of branch targets is not valid, even if at run time it never actually branches there, then the whole instruction will act like a `trap` when executed), and this coupled with verifying that the last instruction of a program is a basic block terminator should make the case of executing an out-of-bounds instruction impossible.

Also could use your opinion [here](https://github.com/gavofyork/graypaper/pull/508#discussion_r2742658974), unless you're fine with how it is right now. (:

davxy commented on text/pvm.tex:
```diff
@@ -33,40 +30,55 @@ \subsection{Basic Definition}
+    }\\
+    \mathfrak{v}_{\text{inst}}\colon&\abracegroup{
+      \tuple{\blob, \bitstring, \pvmreg} &\to \bool \\
+      (\mathbf{c}, \mathbf{k}, \imath) &\mapsto \len{\mathbf{k}} = \len{\mathbf{c}} \land \imath < \len{\mathbf{k}} \land \mathbf{k}_{\imath} = 1 \land \mathbf{c}_\imath \in U
```
Minor nit: this does not affect the overall semantics, but perhaps placing |**k**| = |**c**| at the top level, to be evaluated once, would be more appropriate.

koute replied:
This was deliberate, as on the original line (i.e. the one with `\exists!\,\mathbf{c} ...`) there is no space to fit this, and I didn't want to overcomplicate/uglify the equations, so the easiest thing to do was to just move it to `v_inst` (since, as you've noted, it doesn't matter semantically whether it's there or at toplevel).

But I'd be happy to move it if you can suggest a way to do it nicely. (:

davxy replied:
As mentioned, this does not affect the semantics. If moving it makes the result less readable, keep it as is. This is fine

davxy commented on text/pvm.tex:
```diff
@@ -714,31 +753,32 @@ \subsection{Host Call Definition}
+  &\Psi_H^*\colon \abracegroup{
     \tuple{\begin{aligned}
-      &\blob, \pvmreg, \gas, \regs,\\&\ram, \contextmutator{X}, X
+      &\blob, \pvmreg, \gas, \bool, \regs,\\&\ram, \contextmutator{X}, X
```
```suggestion
      &\blob, \pvmreg, \gas, \bool,\\&\regs, \ram, \contextmutator{X}, X
```

Nit. To shift all of this slightly toward the LHS.

davxy commented on text/pvm.tex:
```diff
@@ -839,3 +879,517 @@ \subsection{Argument Invocation Definition}
+  }{
+    \where \gas¬sim¬state' &= \gas¬sim¬state\text{ except:}\\
+      \gas¬sim¬state'_{\imath} &= \gas¬sim¬state_{\imath} + 1 + \Fskip(\gas¬sim¬state_{\imath}) \\
+      \gas¬sim¬state'_{\remaining¬decodeslots} &= \remaining¬decodeslots - 1 \\
```
```suggestion
      \gas¬sim¬state'_{\remaining¬decodeslots} &= \gas¬sim¬state_{\remaining¬decodeslots} - 1 \\
```

davxy commented on text/pvm.tex:
```diff
@@ -839,3 +879,517 @@ \subsection{Argument Invocation Definition}
+      \gas¬sim¬state_{\imath} + 1 + \Fskip(\gas¬sim¬state_{\imath}) &\otherwise
+    \end{cases}\\
+    \mathbf{c} &\mapsto \instructions \text{ according to \ref{eq:instructions}}\\
+    \gas¬sim¬state'_{\remaining¬decodeslots} &= \remaining¬decodeslots - \cost¬decodeslots(\mathbf{c}, \mathbf{k}, \gas¬sim¬state_{\imath}) \\
```
```suggestion
    \gas¬sim¬state'_{\remaining¬decodeslots} &= \gas¬sim¬state_{\remaining¬decodeslots} - \cost¬decodeslots(\mathbf{c}, \mathbf{k}, \gas¬sim¬state_{\imath}) \\
```

davxy reviewed (dismissed):
Overall, the PR looks good.

Developing a clear mental model of everything in Section A.9 is not exactly straightforward :-)

A few points are not entirely clear to me - for example, the rationale behind the initial values assigned to the simulation state or the meaning of (A, L, S, M, D) letters (ALU, LOAD, STORE, MUL, DIV - I inferred that from the sources, though can be also inferred from the table in A.10...).

koute committed:
Simplify `grow_heap`

koute committed:
Shuffle the newline in `Phi_H^*` to reduce wasted space

koute committed:
Add missing `\gas¬sim¬state`

koute commented:
@davxy Thanks for reviewing again!

Fixed the last few issues; unless I made an extra oopsie somewhere, should be good to go now?

> A few points are not entirely clear to me - for example, the rationale behind the initial values assigned to the simulation state or the meaning of (A, L, S, M, D) letters (ALU, LOAD, STORE, MUL, DIV - I inferred that from the sources, though can be also inferred from the table in A.10...).

Hm, well, I could provide a more detailed rationale for all of the various design choices here, but this is probably out of scope of the GP? (:

But, essentially the answer is almost always either a) because that's the number we've got from measurements and from reading the microarchitectural datasheets of our reference hardware, and/or b) because we think that's a sensible number given the constraints and the target hardware and the recompilation performance we're targeting. (With the caveat that the current set of magic numbers is most likely not final and I will still tweak these before 1.0 as I get better data/try to adversarially break the gas cost model.) For example, the size of the reorder buffer being 32 and having at most 5 instructions start per cycle was very deliberately picked to allow more efficient SIMD-based implementations of the gas cost model while providing good enough predictive power vs having those values be higher (where you start running into very heavy diminishing returns).

zdave-parity commented on text/pvm_invocations.tex:
```diff
@@ -321,21 +321,19 @@ \subsection{General Functions}\label{sec:generalfunctions}
-      % as they want, in which case they will be able to gracefully handle this situation.
-      \tup{\continue, \gascounter - 100, a + c, (\memory_\ram¬access)\interval{a}{h}} &\otherwise \\
-    \end{cases}
+    \using g &= 100 + (\registers_7 - h) \cdot 10 \\
```
Can we just put `g = 100 + \max(\registers_7 - h, 0) \cdot 10` in the left column now, and have gas handled the same way as all the other host calls?

koute replied:
@zdave-parity That wouldn't be equivalent though? The `max` would indeed handle the case of `<= h` but there's still the case of `> b` where that should also be zero for the non-constant part (because the call is no-op). I could duplicate the `ifs` to make it match, but that seems somewhat silly. (Unless we want to charge gas anyway even though no work would be done?)

Alternatively I could get this equation to *return* `g` instead of `\gascounter'`, if you prefer that?

davxy commented on text/pvm.tex:
> limited to being at most 16

If the maximum legal instruction length is 16, why is 24 used as the upper bound in the equation above? Shouldn't it be 15 instead?

koute replied:
The 24 upper bound is for the *skip* function, i.e. the distance to the next instruction.

The "implicit though limited to being at most 16" essentially means "the whole thing is designed so that you only ever need to read at most 16 bytes per instruction to read all of the instruction's operands", or in more practical terms - you can read a single `u128` per instruction and efficiently decode it through a series of bit operations instead of having to read byte-by-byte.

Note: the distance to the next instruction (i.e. the *skip* function) doesn't necessarily have to be the same as the number of bytes used by the opcode + the arguments. (For example, we take advantage of this property to encode a dispatch table at the beginning of JAM blobs to keep the entry points for is_authorized/accumulate/refine at nice round addresses, where the instructions' *skip* is wider than it is necessary.)

If the "However, each instruction [...]" part is confusing I suppose we could delete it? (As it doesn't really hold any normative power per se.)

davxy replied:
> Note: the distance to the next instruction (i.e. the skip function) doesn't necessarily have to be the same as the number of bytes used by the opcode + the arguments. 

Alright. I wrongly assumed that was the case; that is, one immediately follows the other. This clarifies the situation.

davxy commented on text/pvm.tex:
```diff
@@ -749,24 +789,24 @@ \subsection{Host Call Definition}
-  \tup{\mathbf{p}, \mathbf{a}} &\mapsto \begin{cases}
-    \tup{\mathbf{c}, \registers, \mem} &\when \exists! \tup{\mathbf{c}, \mathbf{o}, \mathbf{w}, z, s} \text{ which satisfy equation \ref{eq:conditions}}\\
+  \tup{\jam¬blob, \mathbf{a}} &\mapsto \begin{cases}
+    \tup{\pvm¬blob, \registers, \mem} &\when \exists! \tup{\pvm¬blob, \mathbf{o}, \mathbf{w}, z, s} \text{ which satisfy equation \ref{eq:conditions}}\\
```
When this is rendered, it references equation A.42 only 

<img width="680" height="90" alt="Image" src="https://github.com/user-attachments/assets/340fdf9f-2c73-4807-a106-110c1f4c1df8" />

<img width="762" height="67" alt="Image" src="https://github.com/user-attachments/assets/fb58bae0-6778-4196-8632-05e14794f15c" />

But perhaps A.45 should be referenced as well?

<img width="651" height="34" alt="Image" src="https://github.com/user-attachments/assets/91b8bb54-1bd5-4759-b0df-5742d2aa9856" />

This is strange as looks like the label does indeed encompass both. Perhaps this requires some more LaTeX magic

koute commented on text/pvm_invocations.tex:
```diff
@@ -299,8 +302,43 @@ \subsection{General Functions}\label{sec:generalfunctions}
+    % The total gas cost, based on the number of pages that will become writable after this call.
+    \using g &= 100 + (\registers_7 - h) \cdot 10 \\
+    \tup{\execst', \gascounter', \registers'_7} &\equiv \begin{cases}
+      % Do nothing if the heap is large enough already, or if the requested size was too large.
```
> mmm... Indeed the user can detect a failure by comparing the returned value against what he requested. So perhaps is fine?

Yes, that was the idea - the user can compare the returned value to the requested one to check whether it failed.

davxy commented on text/pvm.tex:
```diff
@@ -839,3 +879,517 @@ \subsection{Argument Invocation Definition}
+\begin{equation}\label{eq:gascostforblock}
+  \gascostforblock\colon \abracegrouptwo{
+    \tup{\blob, \bitstring, \pvmreg} &\to \N\\
+    \tup{\mathbf{c}, \mathbf{k}, \imath} &\mapsto \max(\gas¬sim¬state^{\text{final}}_{\ctr¬cycles} - 3, 1)\\
```
The `max(cycles - 3, 1)` in the final gas cost gives a 3-cycle "free" startup, correct? I'm just curious why?

koute replied:
In a nutshell: programs are very often full of tiny basic blocks, and real CPUs can "see" across those; our gas cost model can't. Considering the rest of the model pretty much assumes the worst case, this optimistically tries to claw back a little bit of gas.

That said, as with the rest of the numbers in the model, this value is subject to change and might be either removed or reduced to zero if it ends up being insecure after more thorough testing.

davxy reviewed (approved):
I think we can go ahead
